File: 자프링-GraphQL-실습/src/main/java/com/junyss/graphqltest/api/equipment/resolver/EquipmentResolver.java
package com.junyss.graphqltest.api.equipment.resolver;

import java.util.List;

import org.springframework.graphql.data.method.annotation.Argument;
import org.springframework.graphql.data.method.annotation.MutationMapping;
import org.springframework.graphql.data.method.annotation.QueryMapping;
import org.springframework.stereotype.Controller;

import com.junyss.graphqltest.api.equipment.model.dto.response.EquipmentsAdvResponseDto;
import com.junyss.graphqltest.api.equipment.service.EquipmentService;
import com.junyss.graphqltest.common.constant.DefaultResponse;
import com.junyss.graphqltest.api.equipment.model.dto.request.EquipmentRequestDto;
import com.junyss.graphqltest.api.equipment.model.dto.response.EquipmentResponseDto;

import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
@Controller // REST API가 아니기 때문에 @Controller 사용
public class EquipmentResolver {

	private final EquipmentService equipmentService;

	/**
	 * @MuationMapping은 REST API 구축 시 @PostMapping과 같은 어노테이션으로 graphql의 Mutaion에 사용.
	 * GraphQL은 Endpoint가 하나이기 때문에 @MutaionMapping만 지정하고, 다른 설정을 해 줄 필요가 없음.
	 */
	@MutationMapping
	// @Argument는 @RequestBody, @RequestParam과 같은 매개 변수 값을 지정할 때 사용.
	public DefaultResponse<String> saveForEquipment (@Argument("input") EquipmentRequestDto equipmentRequestDto) {
		return equipmentService.saveForEquipment(equipmentRequestDto);
	}

	/**
	 * @QueryMapping도 Rest API 구축 시 @GetMapping과 같은 어노테이션이며, @SubscriptionMapping도 존재
	 */
	@QueryMapping
	public DefaultResponse<List<EquipmentResponseDto>> getEquipmentList (
		@Argument String equipmentId,
		@Argument String usedBy,
		@Argument String newOrUsed,
		@Argument Integer page,
		@Argument Integer size) {
		return equipmentService.getEquipmentList(equipmentId, usedBy, newOrUsed, page, size);
	}

	@QueryMapping
	public DefaultResponse<List<EquipmentsAdvResponseDto>> getEquipmentsAdv (
		@Argument String equipmentId,
		@Argument String usedBy,
		@Argument String newOrUsed,
		@Argument Integer page,
		@Argument Integer size) {
		return equipmentService.getEquipmentsAdv(
			equipmentId,
			usedBy,
			newOrUsed,
			page,
			size);
	}

	@QueryMapping
	public DefaultResponse<EquipmentResponseDto> getEquipment (@Argument String equipmentId) {
		return equipmentService.getEquipment(equipmentId);
	}

	@MutationMapping
	public DefaultResponse<String> updateEquipment (@Argument("input") EquipmentRequestDto equipmentRequestDto) {
		return equipmentService.updateEquipment(equipmentRequestDto);
	}

	@MutationMapping
	public DefaultResponse<String> deleteEquipment (@Argument String equipmentId) {
		return equipmentService.deleteEquipment(equipmentId);
	}
}

File: kotlin-graphql-test/src/main/kotlin/com/junyharangstudy/kotlingraphqltest/api/equipment/resolver/EquipmentResolver.kt
package com.junyharangstudy.kotlingraphqltest.api.equipment.resolver

import com.junyharangstudy.kotlingraphqltest.api.common.constant.PageRequestDto
import com.junyharangstudy.kotlingraphqltest.api.equipment.model.dto.request.EquipmentCreateRequestDto
import com.junyharangstudy.kotlingraphqltest.api.equipment.model.dto.request.EquipmentSearchRequestDto
import com.junyharangstudy.kotlingraphqltest.api.equipment.model.dto.request.EquipmentUpdateRequestDto
import com.junyharangstudy.kotlingraphqltest.api.equipment.model.dto.response.EquipmentAdvResponseDto
import com.junyharangstudy.kotlingraphqltest.api.equipment.model.dto.response.EquipmentResponseDto
import com.junyharangstudy.kotlingraphqltest.api.equipment.service.EquipmentService
import com.junyharangstudy.kotlingraphqltest.common.constant.DefaultResponse
import org.springframework.graphql.data.method.annotation.Argument
import org.springframework.graphql.data.method.annotation.MutationMapping
import org.springframework.graphql.data.method.annotation.QueryMapping
import org.springframework.stereotype.Controller

@Controller
class EquipmentResolver (
    private val equipmentService: EquipmentService
) {
    /**
     * @MuationMapping은 REST API 구축 시 @PostMapping과 같은 어노테이션으로 graphql의 Mutaion에 사용.
     * GraphQL은 Endpoint가 하나이기 때문에 @MutaionMapping만 지정하고, 다른 설정을 해 줄 필요가 없음.
     */

    @MutationMapping
    fun saveEquipment(@Argument("input") equipmentCreateRequestDto: EquipmentCreateRequestDto): DefaultResponse<String> {
        return equipmentService.saveEquipment(equipmentCreateRequestDto)
    }

    /**
     * @QueryMapping도 Rest API 구축 시 @GetMapping과 같은 어노테이션이며, @SubscriptionMapping도 존재
     */

    @QueryMapping
    fun getEquipmentList (
        @Argument equipmentSearchRequestDto: EquipmentSearchRequestDto?,
        @Argument pageRequestDto: PageRequestDto?,
    ): DefaultResponse<List<EquipmentResponseDto>> {
        return equipmentService.getEquipmentList(pageRequestDto, equipmentSearchRequestDto)
    }

    @QueryMapping
    fun getEquipmentsAdv (
        @Argument equipmentSearchRequestDto: EquipmentSearchRequestDto?,
        @Argument pageRequestDto: PageRequestDto?
    ): DefaultResponse<List<EquipmentAdvResponseDto>> {
        return equipmentService.getEquipmentAdv(pageRequestDto, equipmentSearchRequestDto)
    }

    @QueryMapping
    fun getEquipment (@Argument equipmentId: String): DefaultResponse<EquipmentResponseDto> {
        return equipmentService.getEquipment(equipmentId)
    }

    @MutationMapping
    fun updateEquipment (
        @Argument("equipmentId") equipmentId: String,
        @Argument("input") equipmentUpdateRequestDto: EquipmentUpdateRequestDto): DefaultResponse<String> {
        return equipmentService.updateEquipment(equipmentId, equipmentUpdateRequestDto)
    }

    @MutationMapping
    fun deleteEquipment (@Argument equipmentId: String): DefaultResponse<String> {
        return equipmentService.deleteEquipment(equipmentId)
    }
}